[shopsys] added documentation about working with forms (#887)
shopsys/shopsys: docs/introduction/using-form-types.md

# Using Form Types
In this article we will show what types you can use when creating or editing forms, what options they have and what they do.

> ## Default Options

Reviewer: 2nd lvl titles are without Capitalization

Every form type has some default options that can be used for various things

### macro
Defaults to `null`.

> This option is used as array with two options (`name`, `recommended_length`) and its currently used for working with seo:

Author: fixed

- `name` - this option has two values that can be used

>     - `seoFormRowMacros` - this value shows single TextType field with information about current length and recommended length

Author: fixed

### display_format
Defaults to `null`.
You can use this option if you want to have no padding in your rendered form_row.
To achieve this you need to fill this option with constant `FormRenderingConfigurationExtension::DISPLAY_FORMAT_MULTIDOMAIN_ROWS_NO_PADDING`

### js_container
Defaults to `null`.
This option is used as array with two options (`container_class`, `image`) and an be used to wrap your form_row in `<div class="{{ js_container.container_class }}" data-type="{{ js_container.data_type }}">`

### is_plugin_data_group
Defaults to `false`.
This option can be set to `true` in empty FormType if you want to add part of form from plugin.

### render_form_row
Defaults to `true`.
If you want to add your FormType and don't render it's `form_label` and `form_errors` you can set this option to false.
This can be sed to create FormTypes that can be used in other FormTypes like the ones that are listed below

## Form Types
We created some form types which can help you with creating your own form types.
Here you can find information about what they do and what options they have

### [GroupType](../../packages/framework/src/Form/GroupType.php)

> Group type is used for creating groups of containers. It is not mapped onto any property and it inherits data into group

Author: fixed

so you can work with forms in this group same as you were before. `GroupType` makes sure to render your fields into nicely

> styled `div` wrap.

Author: fixed

> 

Author: fixed

`GroupType` comes with few options that you can use for even more comfortable work.

> #### `label`

Author: its an option for GroupType

### [DisplayOnlyCustomerType](../../packages/framework/src/Form/DisplayOnlyCustomerType.php)
Displays name of a registered customer along with a link to his/her detail.
If there is no customer set, `unregistered customer` text will be displayed instead.

### [DisplayOnlyType](../../packages/framework/src/Form/DisplayOnlyType.php)
Sometimes form needs to only display information but does not need to change and persist this data, for this usages
there is `DisplayOnlyType` which does not map property onto `entity` and let you to display your own data.

### [DisplayOnlyUrlType](../../packages/framework/src/Form/DisplayOnlyUrlType.php)
For displaying custom URL based on routing system, there can be used `DisplayOnlyUrlType`.

### [LocalizedFullWidthType](../../packages/framework/src/Form/LocalizedFullWidthType.php)
For displaying localized field in vertical order of full-width label and inputs, there can be used `LocalizedFullWidthType`.

### [OrderItemsType](../../packages/framework/src/Form/OrderItemsType.php)
Displays editable table of OrderItems from provided Order.

### [WarningMessageType](../../packages/framework/src/Form/WarningMessageType.php)

> Sometimes the form needs to contain some information that is important for viewer, for this usage there is  `WarningMessageType`

Reviewer: ```suggestion
Displays highlighted message with warning icon.
```

> that shows highlighted message with warning icon.

Reviewer: ```suggestion
```

### [LocalizedType](../../packages/framework/src/Form/Locale/LocalizedType.php)
Compound type that renders one form of given type for each locale.
Returns array indexed by locale.

> #### entry_type

Reviewer: ```suggestion
#### entry_type option
```
what about this way ? if this make sense please do it for other options

Author: BV

Defaults to `TextType::class`.
It’s used to define what FormType should be used for the inner forms.

#### entry_options
An array of options that is used for every inner form.

#### main_constraints

> An array of Constraints that is user for field with same locale as administration.

Reviewer: ```suggestion
An array of constraints that is used for field with the same locale as administration.
```

### [CategoriesType](../../packages/framework/src/Form/CategoriesType.php)
Displays a tree of all categories for given `domain_id` with checkboxes for each category created by `CategoryCheckBoxType` and returns array indexed by checked category ids.

#### domain_id
Required option that defines for what domain should the categories be listed.

### [CategoryCheckboxType](../../packages/framework/src/Form/CategoryCheckboxType.php)

> Creates checkbox with and label with category name if name of the form is same as category id.

Reviewer: ```suggestion
Creates checkbox and label with category name if name of the form is the same as category id.
```

> Adds `visible`, `category_name`, `has_children` and `level` vars from category on domain for given `domain_id` to FormView so you can easily work with the checkbox.

Reviewer: ```suggestion
Adds `visible`, `category_name`, `has_children` and `level` vars from category on domain for given `domain_id` to `FormView` so you can easily work with the checkbox.
```

#### domain_id
Required option that defines from what domain should the category be listed.

### [ColorPickerType](../../packages/framework/src/Form/ColorPickerType.php)
Displays text field with box of given color that shows color picker when clicked.

### [DatePickerType](../../packages/framework/src/Form/DatePickerType.php)
Displays field that shows date picker when clicked.

#### format
Defines in what format should be the date shown.
DatePickerType has 2 constants that can be used:
- FORMAT_PHP = 'dd.MM.yyyy'
- FORMAT_JS = 'dd.mm.yy'

> Defaults to DatePickerType::FORMAT_PHP.

Reviewer: maybe defaults to could be below title since it is like that for `Default options`

### [DisplayOnlyDomainIconType](../../packages/framework/src/Form/DisplayOnlyDomainIconType.php)
Displays domain icon with the domains name for given domain ID in `data` option.

### [DomainsType](../../packages/framework/src/Form/DomainsType.php)

> Displays list non-required checkboxes for every domain with the domain name as label.

Reviewer: ```suggestion
Displays list of non-required checkboxes for every domain with the domain name as label.
```

### [DomainType](../../packages/framework/src/Form/DomainType.php)
Displays select box with all domain names or domain URLs.

#### displayUrl
If you set this option to `true`, domain url will be shown instead of domain name.

> Defaults to false.

Reviewer: switch lines please because of `Default options` look

### [FileUploadType](../../packages/framework/src/Form/FileUploadType.php)

> Displays a widget that let's you to upload files by dragging them to the widget or selecting them from your computer.

Reviewer: ```suggestion
Displays a widget that lets you to upload files by dragging them to the widget or selecting them from your computer.
```

> After a file or files are uploaded it shows box for every file and lets you download the files or delete them.

Reviewer: ```suggestion
After a file or files are uploaded it shows box for every file and lets you to download the files or delete them.
```

#### file_constraints
An array of constraints that should be applied for the uploaded file.

#### multiple

> Boolean option that defines if you are able to upload more than one file

Reviewer: is default of this false?
```suggestion
Boolean option that defines if you are able to upload more than one file.
```

#### info_text

> Required option that needs to be string or null and is a text that is shown under the upload icon

Reviewer: ```suggestion
Required option that needs to be string or null and is a text that is shown under the upload icon.
```

### [FriendlyUrlType](../../packages/framework/src/Form/FriendlyUrlType.php)

> Displays a select box with domain urls and text field that lets you create friendly url on selected domain with your valid slug.

Reviewer: ```suggestion
Displays a select box with domain urls and text field that lets you to create friendly url on selected domain with your valid slug.
```

> Uses `DomainType` to display select box with domain urls

Reviewer: ```suggestion
Uses `DomainType` to display select box with domain urls.
```

If `product` is null shows table with every pricing group for every domain.

#### product
Required option that defines for what product should the prices be listed.

### [ProductsType](../../packages/framework/src/Form/ProductsType.php)
Displays a list of products.
The widget adds button that after clicking opens a popup window that lets you search and pick products except for `main_product`.
After picking you can sort or delete picked products.

#### allow_add
Boolean option that defines if products can be added to the list

#### allow_delete
Boolean option that defines if products can be removed from the list

#### sortable
Boolean option that defines if products in the list can be sorted

#### allow_main_variants
Boolean option that defines if main variants can be added

#### allow_variants
Boolean option that defines if variants can be added

#### main_product
Option that defines for what product are you picking products. `main_product` can't be picked.

#### label_button_add
Option that defines text of button for adding products

### [ProductType](../../packages/framework/src/Form/ProductType.php)

> Displays a widget that lets you pick one product.

Author: added + added to options below

Displays select box with all domain names or domain URLs.

#### displayUrl

> Defaults to false.  

Reviewer: ```suggestion
Defaults to `false`.  
```
